### DISCUSSION TITLE
Got lost adding Bluetooth commands

### DISCUSSION THREAD

**Discussion by JustMe-NL** [OMOTE-Community/OMOTE-Firmware#74] [General]
I want to add some bluetooth keyboard keys to certain hardware keys in a scene but I have some trouble figuring out how this is structured. As far as I can see these commands live in both device_keybload_ble and commandHandler.

But if I want a hardware button to send the key 'i', where do I start to add this command?

**Comment by KlausMu** [comment]
The command is already available. In code (not in the keymap) you could write:
`executeCommand(KEYBOARD_SENDSTRING, "i");`
You could also write
`executeCommand(KEYBOARD_BLE_SENDSTRING, "i");`
which explicitely uses the bluetooth keyboard.
"KEYBOARD_*" (without BLE and MQTT) is only a wrapper and equals to the keyboard which is active.

If you want to add this command to a keymap, it is not directly possible because of the additional payload ("i" in that case). So you have to define a new command.

device_keyboard_ble.h
```
extern uint16_t KEYBOARD_BLE_KEY_I;
```
device_keyboard_ble.cpp
```
uint16_t KEYBOARD_BLE_KEY_I;

  get_uniqueCommandID(&KEYBOARD_BLE_KEY_I)   ;
  ...
  register_command_withID( KEYBOARD_BLE_KEY_I    , makeCommandData(BLE_KEYBOARD, {std::to_string(KEYBOARD_BLE_KEY_I)   }));
  ...
  } else if (command == KEYBOARD_BLE_KEY_I) {
    if (doLog) {Serial.printf("KEY I received\r\n");}
    keyboardBLE_sendString(std::string("I").c_str());
```
I think you will find where you have to put these lines.

Maybe it would be easier to allow additional payload in the keymap ...

Which bluetooth remote do you want to replace which has characters on it?

**Reply by JustMe-NL** [reply]
Thank you, that did the trick. I made the mistake of registering the command twice, once in de device itself and once in the keyboard. But then again, I did forget the .c_str() as well so maybe that send me in the wrong direction.

The device in question is a Nvidea Shield running Kodi and the commands are for use in Kodi. My version of Shield does support some IR commands but not enough to be comfortable.

I previously created an extra gui page with the LVGL keyboard (custom map because otherwise it's to small) so I can rename items in the favorites list in Kodi through Bluetooth.

I have added my smarthome gui to every scene as last the page so next job will be to create a function to map a hardwarebutton to a specific gui screen (if available).
